src/tests/integration/MetaMorpho.t.sol

```diff
@@ -0,0 +1,479 @@
+        uint256 highSupplyAfter = morpho.market(marketParamsHighLltv.id()).totalSupplyAssets;
+
+        uint256 expectedDepositedInLow = capLow;
+        uint256 expectedDepositedInHigh = d3mDeposit - capLow;
+
+        assertEq(lowSupplyAfter, lowSupplyBefore + expectedDepositedInLow, "lowSupplyAfter");
+        assertEq(highSupplyAfter, highSupplyBefore + expectedDepositedInHigh, "highSupplyAfter");
+    }
+
+    function testReallocate(int256 d3mDeposit, uint256 reallocation) public {
+        d3mDeposit = bound(d3mDeposit, 0, type(int256).max);
+
+        setCap(marketParamsHighLltv, type(uint184).max);
+
+        adjustDebt(d3mDeposit);
+
+        uint256 vaultSupplyAssets = pool.assetBalance();
+        reallocation = bound(reallocation, 0, vaultSupplyAssets);
+
+        morpho.accrueInterest(marketParams);
+        morpho.accrueInterest(marketParamsHighLltv);
+        uint256 lowSupplyBefore = morpho.market(marketParams.id()).totalSupplyAssets;
+        uint256 highSupplyBefore = morpho.market(marketParamsHighLltv.id()).totalSupplyAssets;
+
+        // Reallocate from low to high lltv.
+        MarketAllocation[] memory allocations = new MarketAllocation[](2);
+        allocations[0] = MarketAllocation({marketParams: marketParams, assets: vaultSupplyAssets - reallocation});
+        allocations[1] = MarketAllocation({marketParams: marketParamsHighLltv, assets: type(uint256).max});
+        vm.prank(IMetaMorpho(spDai).owner());
+        spDai.reallocate(allocations);
+
+        uint256 lowSupplyAfter = morpho.market(marketParams.id()).totalSupplyAssets;
+        uint256 highSupplyAfter = morpho.market(marketParamsHighLltv.id()).totalSupplyAssets;
+
+        assertEq(lowSupplyAfter, lowSupplyBefore - reallocation);
+        assertEq(highSupplyAfter, highSupplyBefore + reallocation);
+    }
+
+    function testWithdrawLiquid(uint256 target1, uint256 target2) public {
+        target1 = bound(target1, 0, uint256(type(int256).max));
+
+        uint256 marketSupplyStart = morpho.market(marketParams.id()).totalSupplyAssets;
+
+        adjustDebt(int256(target1));
+
+        target2 = bound(target2, 0, pool.assetBalance());
+
+        uint256 marketSupplyMiddle = morpho.market(marketParams.id()).totalSupplyAssets;
+
+        uint256 depositedAssets1 = min(target1, maxLineScaled);
+        uint256 expectedWithdraw = min(pool.maxWithdraw(), pool.assetBalance() - target2);
+        uint256 depositedAssets2 = depositedAssets1 - expectedWithdraw;
+
+        // Set target assets at `target2` and exec.
+        vm.prank(operator);
+        plan.setTargetAssets(target2);
+        hub.exec(ilk);
+
+        uint256 marketSupplyEnd = morpho.market(marketParams.id()).totalSupplyAssets;
+
+        assertEq(marketSupplyEnd, marketSupplyMiddle - expectedWithdraw);
+        assertEq(marketSupplyEnd, marketSupplyStart + depositedAssets2);
+    }
+
+    function testWithdrawIlliquid(uint256 target1, uint256 target2, uint256 borrow) public {
```

Comment:
I'm a bit confused about this test. I'd assume that if we are testing withdraw for an illiquid market is because we need to withdraw more than the pool.maxWithdraw();
Then the following assert should pass always:
```
assertGt(pool.assetBalance() - target2, pool.maxWithdraw());
```
(which it doesn't)
We could even be executing a full run without any illiquid case.

I think we should start bounding the borrow value to be minimum the amount that was already available (marketSupplyStart - marketBorrowStart) and then make sure poolAsset() - target 2 > available funds.

Although I might be understating wrongly the meaning of this test.

Reply:
you are right, fixed in 576c75cb8a76dbbd3390335920307bd3f09b7647

now we try to withdraw more than the market's liquidity

```solidity
target2 = bound(target2, 0, pool.assetBalance() - liquidityMiddle);
```